tephra-core/src/main/java/org/apache/tephra/TransactionManager.java

```diff
@@ -958,25 +983,31 @@ private void doAbort(long writePointer, long[] checkpointWritePointers, Transact
     // makes tx visible (assumes that all operations were rolled back)
     // remove from in-progress set, so that it does not get excluded in the future
     InProgressTx removed = inProgress.remove(writePointer);
+    boolean isInvalid = false;
     if (removed == null) {
       // tx was not in progress! perhaps it timed out and is invalid? try to remove it there.
       if (invalid.rem(writePointer)) {
+        isInvalid = true;
         // remove any invalidated checkpoint pointers
         // this will only be present if the parent write pointer was also invalidated
         if (checkpointWritePointers != null) {
-          for (int i = 0; i < checkpointWritePointers.length; i++) {
-            invalid.rem(checkpointWritePointers[i]);
+          for (long checkpointWritePointer : checkpointWritePointers) {
+            invalid.rem(checkpointWritePointer);
           }
         }
         invalidArray = invalid.toLongArray();
         LOG.info("Tx invalid list: removed aborted tx {}", writePointer);
-        // removed a tx from excludes: must move read pointer
-        moveReadPointerIfNeeded(writePointer);
       }
-    } else {
-      // removed a tx from excludes: must move read pointer
-      moveReadPointerIfNeeded(writePointer);
     }
+    if (!isInvalid) {
```

Comment:
the meaning is actually whether the parent has been invalidated. If so, then the previous if() has removed all children from the invalid set. In all other cases, we have to assume they are still in the inProgress map and remove them. But ok, I can rename it.